[FX][docs] Add limitations of symbolic tracing

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#50638 [FX][docs] Add limitations of symbolic tracing**

![image](https://user-images.githubusercontent.com/4685384/104791737-ae684000-5750-11eb-92e1-15ff95240986.png)
![image](https://user-images.githubusercontent.com/4685384/104791745-b4f6b780-5750-11eb-86fd-1161f30f4071.png)
![image](https://user-images.githubusercontent.com/4685384/104791755-b9bb6b80-5750-11eb-8c95-e9764ccd2446.png)
![image](https://user-images.githubusercontent.com/4685384/104791768-c3dd6a00-5750-11eb-9e15-d0c66ebcdfe2.png)
![image](https://user-images.githubusercontent.com/4685384/104791774-c770f100-5750-11eb-8ceb-bedd1a66d40f.png)


Differential Revision: [D25933780](https://our.internmc.facebook.com/intern/diff/D25933780)

Commit:
[FX][docs] Add limitations of symbolic tracing

[ghstack-poisoned]

vadimkantorov commented:
A question about tracing: is there some analog of `map`/`scan`/`reduce` op that could replace a for loop (and can be similar to sweep algorithms like cumsum or to RNNs)?

jamesr66a commented:
@vadimkantorov at this time, no. We've made the decision for control flow constructs to be out-of-scope in FX since transformations that involve the loop construct itself are relatively rare and supporting control flow has deep implications for the analyzability/transformability for all other use cases. Additionally, functional scan/map/reduce constructs are un-Pythonic and not recommended (see https://www.artima.com/weblogs/viewpost.jsp?thread=98196).

Code that contains these dynamic control structures can be encapsulated into a leaf module with a functional signature and still be used inside an FX-traced model. The only limitation is transformations on the loop itself (e.g. code motion) are not supported at this time.

We can, of course, revisit this decision (reversing the decision is purely additive, after all), but for 1.8 we are sticking to pure basic-blocks.

vadimkantorov commented:
I propose to state this explicitly in the docs: control flow is not supported for now. This is a question that many users would have.

Well, basic Python-icity is not very concerned about potential limitations of compilers / transforms / export / speed of loops. So I would argue for more performance/compilation-oriented code, it makes sense to support these constructs. I created a separate issue about them: https://github.com/pytorch/pytorch/issues/50688

Commit:
Update on "[FX][docs] Add limitations of symbolic tracing"


![image](https://user-images.githubusercontent.com/4685384/104791737-ae684000-5750-11eb-92e1-15ff95240986.png)
![image](https://user-images.githubusercontent.com/4685384/104791745-b4f6b780-5750-11eb-86fd-1161f30f4071.png)
![image](https://user-images.githubusercontent.com/4685384/104791755-b9bb6b80-5750-11eb-8c95-e9764ccd2446.png)
![image](https://user-images.githubusercontent.com/4685384/104791768-c3dd6a00-5750-11eb-9e15-d0c66ebcdfe2.png)
![image](https://user-images.githubusercontent.com/4685384/104791774-c770f100-5750-11eb-8ceb-bedd1a66d40f.png)


Differential Revision: [D25933780](https://our.internmc.facebook.com/intern/diff/D25933780)

[ghstack-poisoned]

Chillee commented on docs/source/fx.rst:
```diff
@@ -10,7 +10,261 @@ Overview
+``MyModule`` with different values for that parameter have different
+code. This is a valid pattern that is supported by symbolic tracing.
+
+Many instances of dynamic control flow are semantically static control
```
Might be worth adding something about the workaround we mentioned previously

Ie:

```
def f(x, flag):
    if flag: return x
    else: return x*2
    
fx.symbolic_trace(f) # Fails!

def g(flag):
    return lambda x: f(x, flag)
    
new_f = g(flag=True)
fx.symbolic_trace(new_f)
```

ansley commented on docs/source/fx.rst:
```diff
@@ -10,7 +10,261 @@ Overview
+Static Control Flow
+~~~~~~~~~~~~~~~~~~~
+
+On the other hand, so called *static control flow* is supported. Static
```
so-called

Commit:
Update on "[FX][docs] Add limitations of symbolic tracing"


![image](https://user-images.githubusercontent.com/4685384/104791737-ae684000-5750-11eb-92e1-15ff95240986.png)
![image](https://user-images.githubusercontent.com/4685384/104791745-b4f6b780-5750-11eb-86fd-1161f30f4071.png)
![image](https://user-images.githubusercontent.com/4685384/104791755-b9bb6b80-5750-11eb-8c95-e9764ccd2446.png)
![image](https://user-images.githubusercontent.com/4685384/104791768-c3dd6a00-5750-11eb-9e15-d0c66ebcdfe2.png)
![image](https://user-images.githubusercontent.com/4685384/104791774-c770f100-5750-11eb-8ceb-bedd1a66d40f.png)


Differential Revision: [D25933780](https://our.internmc.facebook.com/intern/diff/D25933780)

[ghstack-poisoned]

Chillee commented on docs/source/fx.rst:
```diff
@@ -10,7 +10,261 @@ Overview
+        return truediv
+    """
+
+Customizing Tracing with the ``Tracer`` class
```
I think this is a confusing segue. `Customizing Tracer with the tracer class` should be a solution to a limitation, it shouldn't be at the top level.

jamesr66a replied:
I wanted it to be top-level so that Tracer documentation is in one place and any other section of the docs can reference the single source of truth. I can see if i can make these references more clear

Chillee reviewed (approved):
Looks mostly good to me :)

Commit:
Update on "[FX][docs] Add limitations of symbolic tracing"


![image](https://user-images.githubusercontent.com/4685384/104791737-ae684000-5750-11eb-92e1-15ff95240986.png)
![image](https://user-images.githubusercontent.com/4685384/104791745-b4f6b780-5750-11eb-86fd-1161f30f4071.png)
![image](https://user-images.githubusercontent.com/4685384/104791755-b9bb6b80-5750-11eb-8c95-e9764ccd2446.png)
![image](https://user-images.githubusercontent.com/4685384/104791768-c3dd6a00-5750-11eb-9e15-d0c66ebcdfe2.png)
![image](https://user-images.githubusercontent.com/4685384/104791774-c770f100-5750-11eb-8ceb-bedd1a66d40f.png)


Differential Revision: [D25933780](https://our.internmc.facebook.com/intern/diff/D25933780)

[ghstack-poisoned]

Commit:
Update on "[FX][docs] Add limitations of symbolic tracing"


![image](https://user-images.githubusercontent.com/4685384/104791737-ae684000-5750-11eb-92e1-15ff95240986.png)
![image](https://user-images.githubusercontent.com/4685384/104791745-b4f6b780-5750-11eb-86fd-1161f30f4071.png)
![image](https://user-images.githubusercontent.com/4685384/104791755-b9bb6b80-5750-11eb-8c95-e9764ccd2446.png)
![image](https://user-images.githubusercontent.com/4685384/104791768-c3dd6a00-5750-11eb-9e15-d0c66ebcdfe2.png)
![image](https://user-images.githubusercontent.com/4685384/104791774-c770f100-5750-11eb-8ceb-bedd1a66d40f.png)


Differential Revision: [D25933780](https://our.internmc.facebook.com/intern/diff/D25933780)

[ghstack-poisoned]
